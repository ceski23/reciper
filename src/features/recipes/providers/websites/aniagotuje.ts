import aniagotujeIcon from 'assets/images/providers/ania_gotuje.png'
import { extractMicrodata } from 'features/recipes/providers/generic/microdata'
import type { RecipesProvider } from 'features/recipes/providers/scrapper'
import type { Recipe } from 'features/recipes/types'
import { isDefined } from 'lib/utils'
import { getTextFromNode, isElementNode } from 'lib/utils/dom'
import { getColorFromImage } from 'lib/utils/images'

export const aniagotuje: RecipesProvider = {
	name: 'AniaGotuje.pl',
	icon: aniagotujeIcon,
	matcher: /https:\/\/aniagotuje\.pl\/.*/,
	scrape: async doc => {
		const partialRecipe = await extractMicrodata(doc)
		const color = partialRecipe.image ? await getColorFromImage(partialRecipe.image) : undefined
		const name = doc.querySelector('.article-content [itemprop="name"]')?.textContent?.trim()

		return {
			...partialRecipe,
			color,
			name,
			ingredients: parseIngredients(doc) ?? partialRecipe.ingredients,
			instructions: parseInstructions(doc) ?? partialRecipe.instructions,
			gallery: parseImages(doc),
		}
	},
}

const parseImages = (doc: Document) =>
	Array
		.from(doc.querySelectorAll('[itemprop="recipeInstructions"] img'))
		.map(elem => elem.getAttribute('data-src'))
		.filter(isDefined)

const parseIngredients = (doc: Document) => {
	const root = doc.querySelector('[itemprop="recipeInstructions"] > .ads-slot-article + div')
	const ingredientsHeaders = root?.querySelectorAll<HTMLParagraphElement>('.ing-header')

	// Single recipe
	if (ingredientsHeaders?.length === 0) {
		const recipeIngredients = Array
			.from(root?.querySelectorAll('.recipe-ing-list [itemprop="recipeIngredient"]') ?? [])
			.map(elem => elem.textContent?.trim())
			.filter(isDefined)
			.map(text => ({ text }))

		return recipeIngredients.length > 0 ? recipeIngredients : undefined
	}

	// Recipe with sub-recipes
	const recipeIngredients = Array.from(ingredientsHeaders ?? []).reduce<{
		ingredients: Recipe['ingredients']
		group?: string
	}>(({ ingredients, group }, element) => {
		const nextGroup = element.textContent?.trim() ?? group
		const nextIngredients = Array
			.from(element.nextElementSibling?.querySelectorAll('[itemprop="recipeIngredient"]') ?? [])
			.map(elem => elem.textContent?.trim())
			.filter(isDefined)
			.map(text => ({ text, group: nextGroup }))

		return {
			ingredients: ingredients.concat(nextIngredients),
			group: nextGroup,
		}
	}, {
		ingredients: [],
		group: undefined,
	}).ingredients

	return recipeIngredients.length > 0 ? recipeIngredients : undefined
}

const parseInstructions = (doc: Document) => {
	const instructions: Recipe['instructions'] = []
	let currentElement: Node | Element | null = doc.querySelector('[itemprop="recipeInstructions"] div + :is(h3, h2)')
	let header: string | undefined

	while (currentElement !== null && currentElement !== undefined) {
		if (
			isElementNode(currentElement)
			&& ['H1', 'H2', 'H3', 'H4'].includes(currentElement.tagName)
		) {
			header = currentElement.textContent?.trim()
		} else {
			const text = getTextFromNode(currentElement)

			if (text !== null && text.length > 0) {
				instructions.push({ text, group: header })
			}
		}

		currentElement = currentElement.nextSibling
	}

	return instructions.length > 0 ? instructions : undefined
}
