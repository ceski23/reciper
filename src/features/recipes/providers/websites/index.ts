export { aniagotuje } from './aniagotuje'
export { kwestiasmaku } from './kwestiasmaku'
export { wszystkiegoslodkiego } from './wszystkiegoslodkiego'
