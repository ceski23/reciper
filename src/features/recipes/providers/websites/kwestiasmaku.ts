import kwestiasmakuIcon from 'assets/images/providers/kwestia_smaku.png'
import { extractMicrodata } from 'features/recipes/providers/generic/microdata'
import type { RecipesProvider } from 'features/recipes/providers/scrapper'
import type { Recipe } from 'features/recipes/types'
import { isDefined } from 'lib/utils'
import { getTextFromNode } from 'lib/utils/dom'
import { getColorFromImage } from 'lib/utils/images'
import { parseValidNumber } from 'lib/utils/numbers'

export const kwestiasmaku: RecipesProvider = {
	name: 'Kwestia Smaku',
	icon: kwestiasmakuIcon,
	matcher: /https:\/\/www\.kwestiasmaku\.com\/.*/,
	scrape: async doc => {
		const partialRecipe = await extractMicrodata(doc)
		const color = partialRecipe?.image ? await getColorFromImage(partialRecipe.image).catch(() => undefined) : undefined
		const description = doc.querySelector('.field-name-field-uwagi-wstepne')?.textContent?.trim()

		const tags = Array
			.from(doc.querySelectorAll('#node-przepis-full-group-kategorie a'))
			.map(elem => elem.textContent?.toLowerCase())
			.filter(isDefined)

		const servingsElement = doc.querySelector('.field-name-field-ilosc-porcji')
		const servings = parseValidNumber(servingsElement?.textContent)

		return {
			...partialRecipe,
			ingredients: parseIngredients(doc),
			instructions: parseInstructions(doc),
			color,
			description,
			tags,
			servings,
			gallery: parseImages(doc),
		}
	},
}

const parseIngredients = (doc: Document): Recipe['ingredients'] => {
	const ingredientsRoot = doc.querySelector('.field-name-field-skladniki')
	if (!ingredientsRoot) return []

	// Multiple ingredients groups
	if (ingredientsRoot.querySelector('.wyroznione')) {
		return Array.from(ingredientsRoot.children).reduce<{
			ingredients: Recipe['ingredients']
			group?: string
		}>(({ ingredients, group }, element) => {
			const nextGroup = (element.tagName === 'DIV' ? element.textContent?.trim() : undefined) ?? group
			const nextIngredients = Array
				.from(element.querySelectorAll('li'))
				.map(elem => getTextFromNode(elem).trim())
				.filter(isDefined)
				.map(text => ({ text, group: nextGroup }))

			return {
				ingredients: ingredients.concat(nextIngredients),
				group: nextGroup,
			}
		}, {
			ingredients: [],
			group: undefined,
		}).ingredients
	}

	// Signle ingredients group
	const recipeIngredients = Array
		.from(ingredientsRoot.querySelectorAll('ul > li'))
		.map(elem => elem.textContent?.trim())
		.filter(isDefined)
		.map(text => ({ text }))

	return recipeIngredients
}

const parseInstructions = (doc: Document): Recipe['instructions'] => {
	const instructionsRoot = doc.querySelector('.field-name-field-przygotowanie')
	if (!instructionsRoot) return []

	// Multiple instruction groups
	if (instructionsRoot.querySelector('.wyroznione')) {
		return Array.from(instructionsRoot.children).reduce<{
			instructions: Recipe['instructions']
			group?: string
		}>(({ instructions, group }, element) => {
			const nextGroup = (element.tagName === 'DIV' ? element.textContent?.trim() : undefined) ?? group
			const nextInstructions = Array
				.from(element.querySelectorAll('li'))
				.map(elem => getTextFromNode(elem).trim())
				.filter(isDefined)
				.map(text => ({ text, group: nextGroup }))

			return {
				instructions: instructions.concat(nextInstructions),
				group: nextGroup,
			}
		}, {
			instructions: [],
			group: undefined,
		}).instructions
	}

	// Signle instructions group
	return Array
		.from(instructionsRoot.querySelectorAll('ul > li'))
		.map(elem => getTextFromNode(elem).trim())
		.filter(isDefined)
		.map(text => ({ text }))
}

const parseImages = (doc: Document) =>
	Array
		.from(doc.querySelectorAll('.view-zdjecia img'))
		.map(elem => elem.getAttribute('src'))
		.filter(isDefined)
