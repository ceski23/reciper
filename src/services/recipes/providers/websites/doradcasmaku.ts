import icon from 'assets/providers/doradca_smaku.png';

import { Recipe } from 'services/recipes';
import type { Provider, RecipeScrapper } from 'services/recipes/providers';
import jsonldScrapper from 'services/recipes/providers/jsonld';

export const DoradcaSmakuProvider: Provider = (() => {
  const name = 'Doradca smaku';
  const url = 'https://www.doradcasmaku.pl/';

  const scrapper: RecipeScrapper = async (doc) => {
    const data = await jsonldScrapper(doc);

    const recipeImageElem = doc.querySelector('.recipe_item .recipe_gallery .img_wrapper img');
    const image = recipeImageElem?.getAttribute('data-src') ?? undefined;

    const partial: Partial<Recipe> = {
      image,
    };

    return Object.assign(data, partial);
  };

  return {
    name, url, scrapper, icon,
  };
})();
