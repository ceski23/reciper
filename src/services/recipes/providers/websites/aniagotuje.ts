import icon from 'assets/providers/ania_gotuje.png';

import { Recipe, RecipeIngredient, RecipeInstruction } from 'services/recipes';
import type { Provider, RecipeScrapper } from 'services/recipes/providers';
import microdataScrapper from 'services/recipes/providers/microdata';
import { colorExtractor } from 'services/recipes/providers/utils';

import { getTextFromNode } from 'utils/dom';
import { isElementNode, nonNullable } from 'utils/guards';
import appLogger from 'utils/logger';
import { removeEmpty } from 'utils/objects';

const log = appLogger.extend('scrapper:aniagotuje');

export const AniaGotujeProvider: Provider = (() => {
  const name = 'AniaGotuje.pl';
  const url = 'https://aniagotuje.pl/';

  const parseIngredients = (doc: Document) => {
    log('parsing ingredients...');

    const root = doc.querySelector('[itemprop="recipeInstructions"] > .ads-slot-article + div');
    const ingredientsHeaders = root?.querySelectorAll<HTMLParagraphElement>('.ing-header');
    const ingredients: RecipeIngredient[] = [];

    // Single recipe
    if (ingredientsHeaders?.length === 0) {
      log('found single ingredients group');

      const groupItemsElement = root?.querySelectorAll('.recipe-ing-list [itemprop="recipeIngredient"]');
      const groupItems = Array.from(groupItemsElement ?? [])
        .map((item) => item.textContent?.trim())
        .filter(nonNullable)
        .map<RecipeIngredient>((text) => ({ text }));

      ingredients.push(...groupItems);

      return ingredients;
    }

    log('found multiple ingredients groups');

    // Recipe with sub-recipes
    ingredientsHeaders?.forEach((ingHeader) => {
      const groupTitle = ingHeader.textContent?.trim();

      const groupItemsElement = ingHeader.nextElementSibling?.querySelectorAll('[itemprop="recipeIngredient"]');
      const groupItems = Array.from(groupItemsElement ?? [])
        .map((item) => item.textContent?.trim())
        .filter(nonNullable)
        .map<RecipeIngredient>((text) => ({ text, group: groupTitle }));

      ingredients.push(...groupItems);
    });

    log('parsed ingredients:', ingredients);

    return ingredients;
  };

  const parseInstructions = (doc: Document) => {
    log('parsing instructions...');

    const instructions: RecipeInstruction[] = [];
    let currentElement: Node | Element | null = doc.querySelector('[itemprop="recipeInstructions"] .recipe-ing-list + h2');
    let header: string | undefined;

    while (currentElement !== null && currentElement !== undefined) {
      if (
        isElementNode(currentElement)
        && ['H1', 'H2', 'H3', 'H4'].includes(currentElement.tagName)
      ) {
        header = currentElement.textContent?.trim();
      } else {
        const text = getTextFromNode(currentElement);

        if (text !== null && text.length > 0) {
          instructions.push({ text, group: header });
        }
      }

      currentElement = currentElement.nextSibling;
    }

    log('parsed instructions:', instructions);

    return instructions;
  };

  const parseImages = (doc: Document) => {
    log('looking for images');

    const imageElements = doc.querySelectorAll('[itemprop="recipeInstructions"] img');
    const images = Array
      .from(imageElements)
      .map((elem) => elem.getAttribute('data-src'))
      .filter(nonNullable);

    log('found images:', images);

    return images;
  };

  // eslint-disable-next-line @typescript-eslint/no-shadow
  const scrapper: RecipeScrapper = async (doc) => {
    log('scrapping recipe...');

    const data = await microdataScrapper(doc);
    log('data from microdata:', data);

    let color;
    if (data?.image) {
      const palette = await colorExtractor(data.image);
      color = palette.Vibrant?.hex;
    }

    const recipeName = doc.querySelector('.article-content [itemprop="name"]')?.textContent?.trim();

    const partial: Partial<Recipe> = {
      color,
      name: recipeName,
      ingredients: parseIngredients(doc),
      instructions: parseInstructions(doc),
      gallery: parseImages(doc),
    };

    log('additional data:', partial);

    return Object.assign(data, removeEmpty(partial));
  };

  return {
    name, url, scrapper, icon,
  };
})();
