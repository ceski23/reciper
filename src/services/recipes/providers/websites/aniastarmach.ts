/* eslint-disable no-restricted-syntax */
import icon from 'assets/providers/ania_starmach.png';

import { Recipe, RecipeIngredient } from 'services/recipes';
import type { Provider, RecipeScrapper } from 'services/recipes/providers';
import jsonldScrapper from 'services/recipes/providers/jsonld';
import { getColorFromImage } from 'services/recipes/providers/utils';

import { nonNullable } from 'utils/guards';
import { removeEmpty } from 'utils/objects';

export const AniaStarmachProvider: Provider = (() => {
  const name = 'Ania Starmach';
  const url = 'https://aniastarmach.pl/';

  const parseIngredients = (doc: Document) => {
    const ingredientElements = doc.querySelectorAll('.recipe-what-to-buy > ul > li');
    let group: string | undefined;
    const ingredients: RecipeIngredient[] = [];

    for (const element of ingredientElements) {
      if (element.classList.contains('recipe-ingredient-heading')) {
        group = element.textContent?.trim();
      } else {
        const text = element.textContent?.trim();
        if (text !== undefined && text.length > 0) {
          ingredients.push(removeEmpty({ text, group }));
        }
      }
    }

    return ingredients;
  };

  // eslint-disable-next-line @typescript-eslint/no-shadow
  const scrapper: RecipeScrapper = async (doc) => {
    const data = await jsonldScrapper(doc);

    const color = data.image ? await getColorFromImage(data.image) : undefined;

    const tagsElements = doc.querySelectorAll('.tags-wrapper a');
    const tags = Array
      .from(tagsElements)
      .map((elem) => elem.textContent?.toLowerCase())
      .filter(nonNullable);

    const partial: Partial<Recipe> = {
      color,
      tags,
      ingredients: parseIngredients(doc),
    };

    return Object.assign(data, removeEmpty(partial));
  };

  return {
    name, url, scrapper, icon,
  };
})();
