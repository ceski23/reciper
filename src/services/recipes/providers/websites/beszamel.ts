import type { Provider, RecipeScrapper } from 'services/recipes/providers';
import jsonldScrapper from 'services/recipes/providers/jsonld';
import icon from 'assets/provider_icons/beszamel.svg';
import { colorExtractor } from 'services/recipes/providers/utils';

export const BeszamelProvider: Provider = (() => {
  const name = 'Beszamel';
  const url = 'https://beszamel.se.pl/';

  // eslint-disable-next-line @typescript-eslint/no-shadow
  const scrapper: RecipeScrapper = async (doc) => {
    const data = await jsonldScrapper(doc);

    let color;
    if (data?.image) {
      const palette = await colorExtractor(data.image);
      color = palette.Vibrant?.hex;
    }

    return {
      ...data,
      color,
    };
  };

  return {
    name, url, scrapper, icon,
  };
})();
