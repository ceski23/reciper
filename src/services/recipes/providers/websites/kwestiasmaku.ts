/* eslint-disable @typescript-eslint/no-loop-func */
/* eslint-disable @typescript-eslint/brace-style */
/* eslint-disable no-restricted-syntax */
import icon from 'assets/providers/kwestia_smaku.png';

import { Recipe, RecipeIngredient, RecipeInstruction } from 'services/recipes';
import type { Provider, RecipeScrapper } from 'services/recipes/providers';
import microdataScrapper from 'services/recipes/providers/microdata';
import { getColorFromImage } from 'services/recipes/providers/utils';

import { getTextFromNode } from 'utils/dom';
import { nonNullable } from 'utils/guards';
import appLogger from 'utils/logger';
import { removeEmpty } from 'utils/objects';

const log = appLogger.extend('scrapper:kwestiasmaku');

export const KwestiaSmakuProvider: Provider = (() => {
  const name = 'Kwestia Smaku';
  const url = 'https://www.kwestiasmaku.com/';

  const parseIngredients = (doc: Document) => {
    const ingredientsRoot = doc.querySelector('.field-name-field-skladniki');
    if (!ingredientsRoot) return [];

    // Multiple ingredients groups
    if (ingredientsRoot.querySelector('.wyroznione')) {
      let group: string | undefined;
      const ingredients: RecipeIngredient[] = [];

      for (const element of ingredientsRoot.children) {
        if (element.tagName === 'DIV') {
          group = element.textContent?.trim() ?? undefined;
        } else {
          const recipeIngredients = Array
            .from(element.querySelectorAll('li'))
            .map((elem) => elem.textContent?.trim())
            .filter(nonNullable)
            .map<RecipeInstruction>((text) => ({ text, group }));

          ingredients.push(...recipeIngredients);
        }
      }

      return ingredients;
    }

    // Signle ingredients group
    const ingredientsElements = ingredientsRoot.querySelectorAll('ul > li');
    const recipeIngredients = Array
      .from(ingredientsElements)
      .map((elem) => elem.textContent?.trim())
      .filter(nonNullable)
      .map((text) => ({ text }));

    return recipeIngredients;
  };

  const parseInstructions = (doc: Document) => {
    const instructionsRoot = doc.querySelector('.field-name-field-przygotowanie');
    if (!instructionsRoot) return [];

    // Multiple instruction groups
    if (instructionsRoot.querySelector('.wyroznione')) {
      let group: string | undefined;
      const instructions: RecipeInstruction[] = [];

      for (const element of instructionsRoot.children) {
        if (element.tagName === 'DIV') {
          group = element.textContent?.trim() ?? undefined;
        } else {
          const recipeInstructions = Array
            .from(element.querySelectorAll('li'))
            .map((elem) => getTextFromNode(elem).trim())
            .filter(nonNullable)
            .map<RecipeInstruction>((text) => ({ text, group }));

          instructions.push(...recipeInstructions);
        }
      }

      return instructions;
    }

    // Signle instructions group
    const instructionsElements = instructionsRoot.querySelectorAll('ul > li');
    const recipeInstructions = Array
      .from(instructionsElements)
      .map((elem) => getTextFromNode(elem).trim())
      .filter(nonNullable)
      .map((text) => ({ text }));

    return recipeInstructions;
  };

  const parseImages = (doc: Document) => {
    log('looking for images');

    const imageElements = doc.querySelectorAll('.view-zdjecia img');
    const images = Array
      .from(imageElements)
      .map((elem) => elem.getAttribute('src'))
      .filter(nonNullable);

    log('found images:', images);

    return images;
  };

  // eslint-disable-next-line @typescript-eslint/no-shadow
  const scrapper: RecipeScrapper = async (doc) => {
    const microdata = await microdataScrapper(doc);
    const description = doc.querySelector('.field-name-field-uwagi-wstepne')?.textContent?.trim() || undefined;

    const tagsElements = doc.querySelectorAll('#node-przepis-full-group-kategorie a');
    const tags = Array
      .from(tagsElements)
      .map((elem) => elem.textContent?.toLowerCase())
      .filter(nonNullable);

    const servingsElement = doc.querySelector('.field-name-field-ilosc-porcji');
    const servings = servingsElement?.textContent
      ? Number.parseInt(servingsElement.textContent, 10)
      : undefined;

    const color = microdata.image ? await getColorFromImage(microdata.image) : undefined;

    const partial: Partial<Recipe> = {
      ingredients: parseIngredients(doc),
      instructions: parseInstructions(doc),
      color,
      description,
      tags,
      servings,
      gallery: parseImages(doc),
    };

    return Object.assign(microdata, removeEmpty(partial));
  };

  return {
    name, url, scrapper, icon,
  };
})();
