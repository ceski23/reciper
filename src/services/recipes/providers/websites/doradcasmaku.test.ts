// eslint-disable-next-line max-classes-per-file
import { expect } from 'vitest';

import { lemoniada, tartaZKremem } from 'services/recipes/providers/fixtures/doradcasmaku';
import { DoradcaSmakuProvider } from 'services/recipes/providers/websites/doradcasmaku';

vi.stubGlobal('Image', class {
  onload: () => void;

  constructor() {
    this.onload = vi.fn();
    setTimeout(() => this.onload());
  }
});

vi.mock('colorthief', () => ({
  default: class {
    // eslint-disable-next-line class-methods-use-this
    getColor = () => [0, 0, 0];

    getPalette = () => [this.getColor()];
  },
}));

const scrapeRecipe = async (data: string) => {
  const parser = new DOMParser();
  const doc = parser.parseFromString(data, 'text/html');
  const partialRecipe = await DoradcaSmakuProvider.scrapper(doc);
  partialRecipe.id = '1';

  return partialRecipe;
};

describe('scrapper', () => {
  it('should scrape valid recipe', async () => {
    const recipe = await scrapeRecipe(lemoniada);
    expect(recipe).toMatchSnapshot();
  });

  it('should scrape valid recipe with ingredients groups', async () => {
    const recipe = await scrapeRecipe(tartaZKremem);
    expect(recipe).toMatchSnapshot();
  });
});
