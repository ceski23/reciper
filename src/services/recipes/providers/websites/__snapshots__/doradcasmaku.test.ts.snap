// Vitest Snapshot v1

exports[`scrapper > should scrape valid recipe 1`] = `
{
  "color": "#000000",
  "description": "Przepis dla wielbicieli grejpfrutów, czyli cytrusowa lemoniada z sokiem grejpfrutowym i szczyptą egzotycznego kardamonu.",
  "gallery": [
    "https://pliki.doradcasmaku.pl/grejfrutowa-lemoniada-z-kardamonem0-4.jpg",
  ],
  "id": "1",
  "image": "https://pliki.doradcasmaku.pl/grejfrutowa-lemoniada-z-kardamonem0-4.jpg",
  "ingredients": [
    {
      "text": "woda gazowana 300 mililitrów",
    },
    {
      "text": "grejpfrut różowy 2 sztuki",
    },
    {
      "text": "cytryna 1/2 sztuki",
    },
    {
      "text": "Cukier ze skórką cytryny do ciast i deserów Prymat 1 opakowanie",
    },
    {
      "text": "Kardamon mielony Prymat 1 łyżeczka",
    },
    {
      "text": "lód",
    },
  ],
  "instructions": [
    {
      "text": "1. Cukier ze skórką cytryny oraz kardamon zalewamy odrobiną gorącej wody i mieszamy aż cukier się rozpuści.",
    },
    {
      "text": "2. Wyciskamy sok z grejfruta i cytryny. Sprawdź jak wycisnąć sok z cytrusów bez pestek pod linkiem.",
    },
    {
      "text": "3. Mieszamy wyciśnięty sok, wodę z cukrem i kardamonem, a następnie zalewamy gazowaną wodą.",
    },
    {
      "text": "4. Lemoniadę należy schłodzić przed podaniem w lodówce lub dodać kostki lodu.",
    },
  ],
  "name": "Grejpfrutowa lemoniada z kardamonem",
  "prepTime": 15,
  "rating": 5,
  "servings": 1,
  "tags": [
    "sok z cytryny",
    "kardamon mielony prymat",
    "grejpfrut różowy",
    "napoje",
  ],
}
`;

exports[`scrapper > should scrape valid recipe with ingredients groups 1`] = `
{
  "color": "#000000",
  "description": "Kawa + kawałek ciasta, brzmi jak deser idealny! A co powiecie na połączenie kawy w sezonowej tarcie z malinami? Na sam widok tej słodkości i na myśl o niej aż robi się słodko!",
  "gallery": [
    "https://pliki.doradcasmaku.pl/tarta-z-kremem-kawowym-i-malinami0-4.jpg",
    "https://pliki.doradcasmaku.pl/tarta-z-kremem-kawowym-i-malinami1-4.jpg",
    "https://pliki.doradcasmaku.pl/tarta-z-kremem-kawowym-i-malinami2-4.jpg",
  ],
  "id": "1",
  "image": "https://pliki.doradcasmaku.pl/tarta-z-kremem-kawowym-i-malinami0-4.jpg",
  "ingredients": [
    {
      "group": "CIASTO",
      "text": "mąka 150 g",
    },
    {
      "group": "CIASTO",
      "text": "masło 100 g",
    },
    {
      "group": "CIASTO",
      "text": "cukier puder 50 g",
    },
    {
      "group": "CIASTO",
      "text": "żółtko jajek 2",
    },
    {
      "group": "KREM",
      "text": "cukier puder 5 łyżeczek",
    },
    {
      "group": "KREM",
      "text": "kawa rozpuszczalna 2 łyżeczki",
    },
    {
      "group": "KREM",
      "text": "wrzątek 1 łyżka",
    },
    {
      "group": "KREM",
      "text": "śmietana kremówka 400 g",
    },
    {
      "group": "KREM",
      "text": "serek mascarpone 125 g",
    },
    {
      "group": "KREM",
      "text": "maliny świeże 300 g",
    },
  ],
  "instructions": [
    {
      "text": "1. Mąkę siekamy z zimnym masło aż do uzyskania konsystencji kaszy. Dodajemy cukier puder i żółtka, chwilkę siekamy nożem, a później rękoma zagniatamy ciasto. Tworzymy kulę i wkładamy do lodówki (min. 1h).",
    },
    {
      "text": "2. Następnie ciasto rozwałkowujemy na cienki placek i wyklejamy nim formę do tarty (24cm).",
    },
    {
      "text": "3. Ciasto pieczemy około 20 minut w 180 stopniach, aż się zarumienieni. Ciasto można nakłuć widelcem (Dodatkowo można je obciążyć grochem lub ceramicznymi kulkami).",
    },
    {
      "text": "4. Kawę zaparzamy w 1 łyżce wrzątku i studzimy.",
    },
    {
      "text": "5. Kremówkę ubijamy na sztywną pianę, dodajemy ostudzoną kawę i mascarpone, delikatnie całość mieszamy aż do połączenia. Dodajemy cukier puder do smaku.",
    },
    {
      "text": "6. Krem wykładamy na wystudzone ciasto. Na kremie układamy umyte i osuszone maliny.",
    },
    {
      "text": "Ciasto przechowujemy w lodówce. Przed podaniem ciasto oprószamy kawą rozpuszczalną i dekorujemy listkami mięty. Smacznego :)",
    },
  ],
  "name": "Tarta z kremem kawowym i malinami",
  "prepTime": 30,
  "rating": 5,
  "servings": 5,
  "tags": [
    "przepisy na ciasta z kremem",
    "kuchnia polska",
    "tarta z kremem kawowym i malinami",
  ],
}
`;
